### DISCUSSION TITLE
How do I propagate the db instance/fastify to a class?

### DISCUSSION THREAD

**Discussion by officer-rosmarino** [fastify/fastify#3696] [General] [answered]
I'm new to node and fastify and i'm trying to understand how to structure the project

I have a file where I define the routes and the handlers.
The handlers might need to reuse the same functions, so I would like to create a service class with useful functions (i've seen this in [fastify/fastify-example-twitter](https://github.com/fastify/fastify-example-twitter)), but I'm not quite sure how to do this.
This is what i'm done and it doesn't feel right.

**server.js**
```
import Fastify from 'fastify';
import dbConnector from './utils/db.js'
import routes from './routes.js'
import ProductService from './product/service.js';

const fastify = Fastify({logger: true});

fastify.register(dbConnector, {
  client: 'pg',
  connection: {
    host: 'localhost',
    port: 5432,
    user: 'user',
    password: 'pwd',
    database: 'testdb'
  }
});

fastify.decorate('ProductService', new ProductService(fastify)); // <------- is this how I should do it?
fastify.register(routes, { prefix: '/v1' });

const start = async () => {
  try {
    await fastify.listen(3000);
  } catch (err) {
    fastify.log.error(err);
    process.exit(1);
  }
}
start();
```

**./utils/db.js**
```
import fastifyPlugin from 'fastify-plugin'
import knex from 'knex';

async function dbConnector(fastify, options) {
  const handler = await knex(options);
  fastify
    .decorate('knex', handler)
    .addHook('onClose', async (instance) => {
      await instance.knex.destroy();
      delete instance.knex;
    });
}
export default fastifyPlugin(dbConnector);
```

**routes.js**
async function productRoutes(fastify, options) {
  //fastify.get('/', getProductHandler);
  fastify.get('/:id', getOneNicknamesHandler)
}
...
export default nicknameRoutes;

async function getProductHandler(req, res) {
  return await this.ProductService.getOneProduct(req.params.id);
}

**/product/service.js**

```
class ProductService {
  constructor (fastify) {
    this.fastify = fastify; // <---- does this makes sense so that I can use this.fastify.knex later in this class' functions?
  }

  async getOne(_id) {
    try {
      const p= await this.fastify.knex('product')
        .where('id', id);
      return p;
    } catch (e) {
      throw e;
    }
  }
}

export default ProductService;
```

Can anyone point me to the right direction?

**Comment by RafaelGSS** [comment, marked as answer]
This is up to you as a developer to decide. Personally, I would not pass the entire fastify instance once you only need: `fastify.knex`.

Example:

```js
await fastify.register(dbConnector, {
  client: 'pg',
  connection: {
    host: 'localhost',
    port: 5432,
    user: 'user',
    password: 'pwd',
    database: 'testdb'
  }
});

fastify.decorate('ProductService', new ProductService(fastify.knex))
```

**Reply by officer-rosmarino** [reply]
That was pretty easy. I was trying to do what I wrote because it didn't seem to work. Turns our I didn't put await in front of the
`fastify.register(dbConnector, [...]`

Simple as that, thanks a lot
